### ISSUE TITLE
RFC: a skate.emit that awaits "interactive" document.readyState

### ISSUE THREAD

**Issue by marcoms** [skatejs/skatejs#241]
Current behaviour in `skate.emit` is to fire the event whenever the function is called. What I propose is to have it check `document.readyState` for `"interactive"` and if true, fire the event, else listen for `DOMContentLoaded` and fire it then.

I suggest this because in many cases where the user listens for an event from a custom element, events fired before the user's listeners can be attached may be fired, potentially breaking the relevant parts of the application. Awaiting `DOMContentLoaded` (apparently the event fired when `document.readyState` becomes `"interactive"` [0]), will allow for user content to load before firing events.

0: https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState#Examples

**Comment by treshugart** [comment]
I think your suggestion here and in #242 are very valuable. We need to do something like this in order to make the lifecycle more predictable. The only downside I can see with this is if someone were to load a component definition and immediately after tried creating / interacting with an element that would be upgraded by that component, they wouldn't be able to interact with it as expected if the document isn't "interactive" yet. Maybe that's ok.

**Comment by marcoms** [comment]
>if someone were to load a component definition and immediately after tried creating / interacting with an element that would be upgraded by that component, they wouldn't be able to interact with it as expected if the document isn't "interactive" yet.

From within the component (`created` et al) or outside?

**Comment by treshugart** [comment]
From outside. For example:

```
<my-component></my-component>
<script src="skate.js"></script>
<script>
  skate('my-component', {
    created: function () {
      this.test = true;
    }
  });

  // false
  document.querySelector('my-component').test;
  skate.create('my-component').test;

  document.addEventListener('DOMContentLoaded', function () {
    // true
    document.querySelector('my-component').test;
    skate.create('my-component').test;
  });
</script>
```

**Comment by marcoms** [comment]
Right I see. Should this work if property getters also await `DOMContentLoaded`?

Isn't this problem more related to #242, though?

**Comment by treshugart** [comment]
I was thinking that if we solve this for #242 that we wouldn't have to worry about this, but I think I'm seeing your use case here. Are you talking about handler's attached to a component from outside a component?

**Comment by marcoms** [comment]
Never mind, I just realised that was me being silly. I thought that with the following code:

```js
properties: {
  test: {
    get() {
      skate.ready(() => this.test);
    }
  }
}
```

`this.test` would be returned from the anonymous function passed to `skate.ready` back to the outside world. Of course, this wouldn't work.
Promises could be used to get a similar result, but I don't think that there would be any solution that wouldn't change the way that properties are accessed.

**Comment by treshugart** [comment]
Doing a hack day to spike lifecycle consistency tomorrow. Hopefully we'll have a resolution for this.

**Comment by treshugart** [comment]
Redundant because of #258.

**Comment by marcoms** [comment]
I don't think this in particular was addressed by #258, which deals with event bubbling rather than delaying the event
